File: sdk/typescript/src/builder.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
/**
 * `AgentContext` construction (§4).
 *
 * Stateful per-session builder that owns the L0 envelope
 * (agent/session/sequence) and exposes one method per interception
 * point that fills L1 and sets `target`.
 */

import { AgentContext, InterceptionPoint, JsonValue, SPEC_VERSION } from "./index";

function now(): string {
  return new Date().toISOString();
}

export class AgentContextBuilder {
  private seq = 0;
  private readonly agent: { id: string; framework: string; name?: string; version?: string };
  private readonly session: { id: string; started_at?: string };

  constructor(opts: {
    agentId: string;
    framework: string;
    sessionId: string;
    agentName?: string;
    agentVersion?: string;
    sessionStartedAt?: string;
  }) {
    this.agent = { id: opts.agentId, framework: opts.framework };
    if (opts.agentName) this.agent.name = opts.agentName;
    if (opts.agentVersion) this.agent.version = opts.agentVersion;
    this.session = { id: opts.sessionId };
    if (opts.sessionStartedAt) this.session.started_at = opts.sessionStartedAt;
  }

  private envelope(ip: InterceptionPoint, target: JsonValue): AgentContext {
    const ctx: AgentContext = {
      spec: SPEC_VERSION,
      interception_point: ip,
      timestamp: now(),
      sequence: this.seq++,
      agent: { ...this.agent },
      session: { ...this.session },
      target,
    };
    return ctx;
  }

  agentStartup(toolsRegistered: string[]): AgentContext {
    const agentInit = { tools_registered: [...toolsRegistered] };
    const ctx = this.envelope(InterceptionPoint.AgentStartup, agentInit);
    ctx.agent_init = agentInit;
    return ctx;
  }

  input(content: JsonValue, role: "user" | "system" | "external" = "user"): AgentContext {
    const inp = { content, role };
    const ctx = this.envelope(InterceptionPoint.Input, inp);
    ctx.input = inp;
    return ctx;
  }

  preModelCall(modelId: string, messages: Array<{ role: string; content: JsonValue }>): AgentContext {
    const ctx = this.envelope(InterceptionPoint.PreModelCall, messages);
    ctx.model = { id: modelId };
    ctx.messages = messages;
    return ctx;
  }

  postModelCall(
    modelId: string,
    content: JsonValue,
    toolCalls: Array<{ id: string; name: string; args: Record<string, JsonValue> }>,
    finishReason: string,
  ): AgentContext {
    const response = { content, tool_calls: toolCalls, finish_reason: finishReason };
    const ctx = this.envelope(InterceptionPoint.PostModelCall, response);
    ctx.model = { id: modelId };
    ctx.response = response;
    return ctx;
  }

  preToolCall(callId: string, name: string, args: Record<string, JsonValue>): AgentContext {
    const tc = { id: callId, name, args };
    const ctx = this.envelope(InterceptionPoint.PreToolCall, args);
    ctx.tool_call = tc;
    return ctx;
  }

  postToolCall(
    callId: string,
    name: string,
    args: Record<string, JsonValue>,
    value: JsonValue,
    isError = false,
  ): AgentContext {
    const tr = { value, is_error: isError };
    const ctx = this.envelope(InterceptionPoint.PostToolCall, value);
    ctx.tool_call = { id: callId, name, args };
    ctx.tool_result = tr;
    return ctx;
  }

  output(content: JsonValue): AgentContext {
    const out = { content };
    const ctx = this.envelope(InterceptionPoint.Output, out);
    ctx.output = out;
    return ctx;
  }

  agentShutdown(reason: "completed" | "error" | "cancelled"): AgentContext {
    const summary = { reason };
    const ctx = this.envelope(InterceptionPoint.AgentShutdown, summary);
    ctx.summary = summary;
    return ctx;
  }
}

File: sdk/go/agenthooks/builder.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package agenthooks

// AgentContext construction (§4).
//
// AgentContextBuilder owns the L0 envelope (agent/session/sequence) and
// exposes one method per interception point that fills L1 and sets
// target. One instance per session. Port of
// sdk/python/python/agent_hooks/context.py.

import "time"

// AgentContextBuilder is the stateful per-session builder for
// AgentContext values.
type AgentContextBuilder struct {
	agent   map[string]any
	session map[string]any
	seq     int64 // assigned via atomic.AddInt64 (§12.2.3)
	l2      map[string]any
}

// NewAgentContextBuilder constructs a builder with the required L0
// agent/session identifiers.
func NewAgentContextBuilder(agentID, framework, sessionID string) *AgentContextBuilder {
	return &AgentContextBuilder{
		agent:   map[string]any{"id": agentID, "framework": framework},
		session: map[string]any{"id": sessionID},
		l2:      map[string]any{},
	}
}

// WithAgent sets optional L2 agent.name/version.
func (b *AgentContextBuilder) WithAgent(name, version string) *AgentContextBuilder {
	if name != "" {
		b.agent["name"] = name
	}
	if version != "" {
		b.agent["version"] = version
	}
	return b
}

// WithL2 attaches L2 fields (trace, tenant, budgets, actor, ...) to
// every subsequent context.
func (b *AgentContextBuilder) WithL2(fields map[string]any) *AgentContextBuilder {
	for k, v := range fields {
		if v != nil {
			b.l2[k] = v
		}
	}
	return b
}

func nowRFC3339() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000000Z")
}

func cloneMap(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

func (b *AgentContextBuilder) envelope(ip InterceptionPoint, target any) AgentContext {
	ctx := AgentContext{
		"spec":               SpecVersion,
		"interception_point": string(ip),
		"timestamp":          nowRFC3339(),
		"sequence":           atomic.AddInt64(&b.seq, 1) - 1,
		"agent":              cloneMap(b.agent),
		"session":            cloneMap(b.session),
		"target":             target,
	}
	for k, v := range b.l2 {
		ctx[k] = v
	}
	return ctx
}

// AgentStartup builds the agent_startup context (§4.2).
func (b *AgentContextBuilder) AgentStartup(toolsRegistered []string) AgentContext {
	agentInit := map[string]any{"tools_registered": anySlice(toolsRegistered)}
	ctx := b.envelope(AgentStartup, agentInit)
	ctx["agent_init"] = agentInit
	return ctx
}

// Input builds the input context (§4.2).
func (b *AgentContextBuilder) Input(content any, role string) AgentContext {
	inp := map[string]any{"content": content, "role": role}
	ctx := b.envelope(Input, inp)
	ctx["input"] = inp
	return ctx
}

// PreModelCall builds the pre_model_call context (§4.2).
func (b *AgentContextBuilder) PreModelCall(modelID string, messages []map[string]any) AgentContext {
	ctx := b.envelope(PreModelCall, anySlice(messages))
	ctx["model"] = map[string]any{"id": modelID}
	ctx["messages"] = anySlice(messages)
	return ctx
}

// PostModelCall builds the post_model_call context (§4.2).
func (b *AgentContextBuilder) PostModelCall(
	modelID string,
	content any,
	toolCalls []map[string]any,
	finishReason string,
) AgentContext {
	response := map[string]any{
		"content":       content,
		"tool_calls":    anySlice(toolCalls),
		"finish_reason": finishReason,
	}
	ctx := b.envelope(PostModelCall, response)
	ctx["model"] = map[string]any{"id": modelID}
	ctx["response"] = response
	return ctx
}

// PreToolCall builds the pre_tool_call context (§4.2).
func (b *AgentContextBuilder) PreToolCall(callID, name string, args map[string]any) AgentContext {
	tc := map[string]any{"id": callID, "name": name, "args": args}
	ctx := b.envelope(PreToolCall, args)
	ctx["tool_call"] = tc
	return ctx
}

// PostToolCall builds the post_tool_call context (§4.2).
func (b *AgentContextBuilder) PostToolCall(
	callID, name string,
	args map[string]any,
	value any,
	isError bool,
) AgentContext {
	tr := map[string]any{"value": value, "is_error": isError}
	ctx := b.envelope(PostToolCall, value)
	ctx["tool_call"] = map[string]any{"id": callID, "name": name, "args": args}
	ctx["tool_result"] = tr
	return ctx
}

// Output builds the output context (§4.2).
func (b *AgentContextBuilder) Output(content any) AgentContext {
	out := map[string]any{"content": content}
	ctx := b.envelope(Output, out)
	ctx["output"] = out
	return ctx
}

// AgentShutdown builds the agent_shutdown context (§4.2).
func (b *AgentContextBuilder) AgentShutdown(reason string) AgentContext {
	summary := map[string]any{"reason": reason}
	ctx := b.envelope(AgentShutdown, summary)
	ctx["summary"] = summary
	return ctx
}

// anySlice converts []T to []any so encoding/json emits a JSON array
// rather than a base64 string (which it would for []byte) and so the
// AgentContext round-trips as the schema expects.
func anySlice[T any](in []T) []any {
	out := make([]any, len(in))
	for i, v := range in {
		out[i] = v
	}
	return out
}
